File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/insights/VineyardInsightsSingleFlightCoordinator.kt
package com.rork.vinetrack.data.insights

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/** Serializes full Insights sync passes independently for each vineyard. */
class VineyardInsightsSingleFlightCoordinator {
    private data class State(var isRunning: Boolean = false, var needsAnotherPass: Boolean = false)

    private val mutex = Mutex()
    private val states = mutableMapOf<String, State>()

    suspend fun request(vineyardId: String, runPass: suspend () -> Unit) {
        val shouldRun = mutex.withLock {
            val state = states.getOrPut(vineyardId) { State() }
            if (state.isRunning) {
                state.needsAnotherPass = true
                false
            } else {
                state.isRunning = true
                true
            }
        }
        if (!shouldRun) return

        while (true) {
            val failure = runCatching { runPass() }.exceptionOrNull()
            val continueRunning = mutex.withLock {
                val state = states.getValue(vineyardId)
                if (state.needsAnotherPass) {
                    state.needsAnotherPass = false
                    true
                } else {
                    state.isRunning = false
                    states.remove(vineyardId)
                    false
                }
            }
            if (!continueRunning) {
                if (failure != null) throw failure
                return
            }
        }
    }
}

File: ios/VineTrack/App/Insights/VineyardInsightsSingleFlightCoordinator.swift
import Foundation

/// Serializes full Insights sync passes independently for each vineyard.
@MainActor
final class VineyardInsightsSingleFlightCoordinator {
    private struct State {
        var isRunning = false
        var needsAnotherPass = false
    }

    private var states: [UUID: State] = [:]
    private var generation = 0

    var hasRunningPass: Bool { states.values.contains { $0.isRunning } }

    func request(vineyardID: UUID, runPass: @escaping () async -> Void) async {
        let requestGeneration = generation
        if states[vineyardID]?.isRunning == true {
            states[vineyardID]?.needsAnotherPass = true
            return
        }
        states[vineyardID] = State(isRunning: true, needsAnotherPass: false)

        repeat {
            guard requestGeneration == generation else { return }
            states[vineyardID]?.needsAnotherPass = false
            await runPass()
        } while requestGeneration == generation && states[vineyardID]?.needsAnotherPass == true

        guard requestGeneration == generation else { return }
        states[vineyardID] = nil
    }

    func invalidateAll() {
        generation += 1
        states.removeAll()
    }
}
